paddle/fluid/framework/trainer.h

```diff
@@ -91,5 +91,47 @@ class DistMultiTrainer : public MultiTrainer {
   std::shared_ptr<paddle::framework::PullDenseWorker> pull_dense_worker_;
 };
 
+
+class PipelineTrainer : public TrainerBase {
+ public:
+  PipelineTrainer() {}
+  virtual ~PipelineTrainer() {}
+  virtual void Initialize(const TrainerDesc& trainer_desc, Dataset* data_set) override;
+  virtual void InitTrainerEnv(const ProgramDesc& main_program,
+                              const platform::Place& place) override;
+  virtual void InitOtherEnv(const ProgramDesc& main_program) override {}
+  virtual void Run() override;
+  virtual void Finalize() override;
+
+ protected:
+  int rank_num_; // number device to sync parameter
+  int line_num_; // number of pipelines
+  int scope_queue_size_;
+  int sync_steps_;
+  int section_num_;
+  PipelineWorkerParameter pipeline_config_;
+  
+  std::vector<std::unique_ptr<std::vector<std::string>>> in_var_names_;
+  std::vector<std::unique_ptr<std::vector<std::string>>> out_var_names_;
+
+  std::vector<std::vector<std::unique_ptr<std::mutex>>> worker_count_mutex_;
+  std::vector<std::vector<int*>> worker_count_;
+
+  // worker: [section_id][line_id][id]
+  std::vector<std::vector<std::vector<std::shared_ptr<paddle::framework::DeviceWorker>>>> workers_;
+  std::vector<std::thread> section_threads_;
+  
+  std::vector<std::vector<std::unique_ptr<ScopeQueue>>> scope_queues_;
```

Comment:
scope_queues_ should be documented clearly, this is a very important data structure in your design.